### PR TITLE
check tensor rank >= 2 on construction

### PR DESCRIPTION
**Pull request by migwellian**
Added a new method in `TensorShapeValidation` and call it from each Tensor ctor so we catch these issues earlier.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java [hunk 1]
```diff
@@ -147,5 +149,8 @@ public static void checkIndexIsValid(int[] shape, int... index) {
         }
     }
 
+    public static void checkRankIsAtLeastTwo(int[] shape) {
+        Preconditions.checkArgument(shape.length > 1, "Tensors must have rank >=2 : " + Arrays.toString(shape));
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java]: I didn't know we could do this, nice!

File: keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java [hunk 2]
```diff
@@ -84,10 +85,11 @@ public static Nd4jDoubleTensor arange(double start, double end, double stepSize)
     private INDArray tensor;
 
     public Nd4jDoubleTensor(double[] data, int[] shape) {
-        this.tensor = TypedINDArrayFactory.create(data, shape, BUFFER_TYPE);
+        this(TypedINDArrayFactory.create(data, shape, BUFFER_TYPE));
     }
 
     public Nd4jDoubleTensor(INDArray tensor) {
+        TensorShapeValidation.checkRankIsAtLeastTwo(tensor.shape());
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: It's great that we're doing this now, takes the human out of making sure we don't do this (and end-users are informed that things will break). Should we have tests to validate that this is working as expected?

Comment (migwellian) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: good call - added

File: keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java [hunk 3]
```diff
@@ -84,10 +85,11 @@ public static Nd4jDoubleTensor arange(double start, double end, double stepSize)
     private INDArray tensor;
 
     public Nd4jDoubleTensor(double[] data, int[] shape) {
-        this.tensor = TypedINDArrayFactory.create(data, shape, BUFFER_TYPE);
+        this(TypedINDArrayFactory.create(data, shape, BUFFER_TYPE));
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: Does ND4J throw an exception if the shape is zero length? If it doesn't throw an exception and returns a 1x1 then we might want to do the check before the ND4J INDArray creation.

Comment (migwellian) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: ND4J doesn't throw, it returns an INDArray whose shape is an empty `int[]`.

**Pull request merged**